### DISCUSSION TITLE
How the `neetbox.integrations` are designed

### DISCUSSION THREAD

**Discussion by visualDust** [visualDust/neetbox#79] [Ideas]
The problem is, as we want a not-dependency-intensive neetbox, we are trying to remove codes with heavy dependencies from core functions. To be specific, in order to avoid installing package 'thop'(which depends on torch) for non-torch users, we have removed `import torch` from the headings of each python file into the line where the packages are required to overcome the import errors for most of the packages. For example, in `neetbox.torch.profile`:
```python
from thop import profile as _profile
def profile():
    case 1: ... return
    case 2: ... return
    case that needs thop: ... return
```
was changed to:
```python
def profile():
    case 1: ... return
    case 2: ... return
    case that needs thop: 
            assert pkg.is_installed('thop')
            from thop import profile as _profile
            ... 
            return
```
It was simple until we added `neetbox.integrations`. I'm not a pro for Python things, but I found it much more clear if we move things related to third parties into `neetbox.integrations`. So what does the `neetbox.integrations` stand for? In general, `neetbox.integrations` contains plugins for the original neetbox codes which empowers neetbox with third-party-compbilities. 

I first mentioned the idea in #19 , in which I suggest that we should move 'nvidia-smi' and 'conda' command line getters out of `neetbox.utils`. However, since I'm a little confused about how to effectively add `plugins` into python class, I'm not sure what to do next to form the general framework for `neetbox.integrations`. Anyway, I think I need some help. Further discussions and methods are welcomed and appreciated.

**Comment by PommesPeter** [comment]
I think we can use `modulefinder` a builtin python module to integrate `nvidia-smi` or `conda` like command, and achieve the goal above. here is the doc. [modulefinder](https://docs.python.org/zh-cn/3/library/modulefinder.html)

**Comment by visualDust** [comment]
Currently, I'm trying with [patch](https://github.com/visualDust/neetbox/blob/bbfd847a9ec9883423ab782aa6e1abd54a683e6a/neetbox/utils/utils.py#L63) as a workaround. But I think it's not gonna be so easy. I will put my new ideas here as soon I get them.

Not to mention, an example here:
```python
# package neetbox.integrations.tensorboard
from neetbox.logging.logger import Logger
from neetbox.utils.utils import patch
from neetbox.integrations import pkg
from typing import Any

assert pkg.is_installed('tensorboard',try_install_if_not=True)
import tensorboard

@patch
def write_tensorboard(self:Logger, what:Any,):
    pass # to write something into tensorboard
```
The code adds a new function caller 'write_tensorboard' into class Logger. So later in my code, I can do such thing:
```python
from neetbox.logging import logger
import neetbox.integrations

logger.write_tensorboard(whatever)
```

**Comment by visualDust** [comment]
It's a kinda solution, but idk how my IDEs perform with the code completion... I'll find it out later when I have time.

**Comment by visualDust** [comment]
Another example for `patch` in `neetbox.utils.utils`:

![image](https://user-images.githubusercontent.com/33346934/231200439-6c5937f7-d6fb-4817-8d6f-82a7ac67ee68.png)

![image](https://user-images.githubusercontent.com/33346934/231200559-179c07aa-1e76-4d7e-8c9f-4e06047d19a2.png)

**Comment by visualDust** [comment]
@PaperCube @PommesPeter  What do you think? Do you find it plausible?

**Comment by visualDust** [comment]
I'm running out of ideas. I'm rubbish indeed.

**Comment by PaperCube** [comment]
I think `@patch` is a nice design. Nevertheless, I think traditional OOP abstract classes might also be a feasible solution.

**Comment by visualDust** [comment]
another approach scans all submodules in `neetbox.integrations` and let them register a self loading function after workspace is loaded. #72 
https://github.com/visualDust/neetbox/pull/72/commits/cbcc66d67040e7cd6239ce2a494e2f1bd6680ce7

it's kinda piece of scaffold, I will make it into a stronger implementation later, just keep you guys up to date.

**Comment by PommesPeter** [comment]
So, how should we design the `neetbox.intergrations`.

**Comment by visualDust** [comment]
> So, how should we design the `neetbox.intergrations`.

`neetbox.integrations` is now moved to `neetbox.extension`. it automatically iterate submodules and try to initialize them. `neetbox.extension` provides a decorator which allows functions run on neetbox workspace initialization. for example, in `neetbox/extension/environment/hardware.py`:

```python
from neetbox.extension import on_workspace_load

# run this function on workspace initialization
@on_workspace_load(name="hardware-monit")
def load_monit_hardware():
    cfg = get_module_level_config()
    if cfg["monit"]:  # if do monit hardware
        hardware.set_update_intervel(cfg["interval"])

        # watch updates in daemon
        @watch(name="hardware", _channel=SYSTEM_CHANNEL, interval=cfg["interval"], initiative=False)
        def update_env_stat():
            return hardware.json()
```

once after neetbox loads up its config, it runs this function. Therefore, you can add custom sub-modules as extensions in `neetbox.extension`, and specify what to do after workspace config is loaded. note that do not use `get_module_level_config` or any config relating things before neetbox finish reading workspace config, that is, make sure you are using `get_module_level_config()` only in function decorated by `@on_workspace_load`.

**Reply by visualDust** [reply]
this is a temporary workaround which will be redesigned in the next beta release.

**Comment by visualDust** [comment]
# Important updates

in #76, there is a patch fix (https://github.com/visualDust/neetbox/pull/76/commits/ef5b792d02b3468e62d9437cb7c56a3a36336e95) which allows submodules in neetbox.extension to provide default config. Example:

in `neetbox\extension\environment\hardware.py`:

```python
from neetbox.config import export_default_config, get_module_level_config
from neetbox.extension import on_workspace_loaded

...

@export_default_config()
def return_default_config() -> dict:
    return {"monit": True, "interval": 0.5}


# watch updates in daemon
@on_workspace_loaded(name="hardware-monit")
def load_monit_hardware():
    cfg = get_module_level_config()
    if cfg["monit"]:  # if do monit hardware
        hardware.set_update_intervel(cfg["interval"])

        # watch updates in daemon
        @watch(name="hardware", _channel=SYSTEM_CHANNEL, interval=cfg["interval"], initiative=False)
        def update_env_stat():
            return hardware.json()
```

in code above, `@export_default_config()` means that neetbox will take the returned value of the decorated function as default configuration for this module(`neetbox.extension.environment.hardware`).  when user initialize a new neetbox workspace via `neet init`, the provided config will be written into user's config file. config generated for code above:

```toml
[extension.environment.hardware]
monit = true
interval = 0.5
```

plus, `@on_workspace_loaded()` works as we have mentioned before, but the name is changed from `@on_workspace_load` to `@on_workspace_loaded`.
